### ISSUE TITLE
Support for stream module

### ISSUE THREAD

**Issue by sjawhar** [jwilder/nginx-proxy#792]
I wanted to float an idea: support for streaming connections. I put together a quick proof of concept on [this fork](https://github.com/sjawhar/nginx-proxy/tree/stream#streaming-backends). Run `docker-compose -f docker-compose-separate-containers.yml up -d --build` to see connections proxied to both the `whoami` service and a mongo service. You can grab a mongo admin UI and connect to it using mongo.local:27017.

So, what do people think? Is there a valuable use case for this? 

Downsides:
* You can't have multiple streaming connections on the same port

Work left:
* Get it working when docker-gen and nginx are running in the same container
* Block http access over the streaming port

**Comment by clobber88** [comment]
Would this allow me to run OpenVPN behind nginx-proxy?

**Comment by aackerman** [comment]
I'm very interested in this functionality, but I think this would require deviating from the official nginx docker image. They don't seem to provide the stream modules from the official images.

**Comment by sjawhar** [comment]
Yes, it requires building on the official nginx image. You can see the Dockerfile I put together to do that here: https://github.com/sjawhar/nginx-proxy/blob/stream/Dockerfile.nginx

My main concern, as the person who opened this issue, is that I'm not sure it's even a good idea to start with. Maybe it's better to just use HAProxy or something.

**Comment by cgraefe** [comment]
I would find this useful. I'm using the Let's Encrypt companion container to manage SSL certificates for a bunch of HTTP hosts. Now I would like to put my MQTT server behind the proxy and have its certificate renewed automatically.

**Comment by lloydzhou** [comment]
i think, can split `nginx.tmpl` in two files:
`upstream.tmpl` --> `upstream.conf`
`server.tmpl` --> `default.conf`

then, can include `upstream.conf` in `http` and `stream`

**Comment by krystalcode** [comment]
+1 this could be also used to proxy a TURN server.
